pkg/reconciler/jobsource/jobsource.go

```diff
@@ -103,19 +126,119 @@ func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
 
-	// TODO(spencer-p) implement this method
+	// Having a sink is a prereq for starting the job, so we reconcile the sink first
+	if err := r.reconcileSink(ctx, js); err == errSinkInvalid {
+		// No real error, but sink was bad
+		logger.Warn("Sink resolved to a bad URI")
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	if err := r.reconcileJob(ctx, js); err != nil {
+		return err
+	}
 
 	js.Status.ObservedGeneration = js.Generation
 	return nil
 }
 
+// reconcileJob enforces the creation and lifecycle of the job. It assumes that the sink exists and is valid.
+func (r *Reconciler) reconcileJob(ctx context.Context, js *v1alpha1.JobSource) error {
+	job, err := r.getJob(ctx, js, labels.SelectorFromSet(resources.Labels(js)))
+
+	if apierrs.IsNotFound(err) {
+		// No job, must create it
+		job = resources.MakeJob(resources.Arguments{
+			Owner:        js,
+			Namespace:    js.Namespace,
+			Template:     js.Spec.Template,
+			SinkURI:      js.Status.SinkURI,
+			OutputFormat: js.Spec.OutputFormat,
+		})
+
+		job, err := r.KubeClientSet.BatchV1().Jobs(js.Namespace).Create(job)
+		if err != nil || job == nil {
+			msg := "Failed to make Job."
+			if err != nil {
+				msg = msg + " " + err.Error()
+			}
+			js.Status.MarkJobFailed("FailedCreate", msg)
+			return fmt.Errorf("failed to create Job: %s", err)
+		}
+
+		js.Status.MarkJobRunning("Created Job %q.", job.Name)
+		return nil
+	} else if err != nil {
+		js.Status.MarkJobFailed("FailedGet", err.Error())
+		return fmt.Errorf("failed to get Job: %s", err)
+	}
+
+	// Job exists, check if it is done
+	if cond := getJobCompletedCondition(job); cond != nil && jobConditionSucceeded(cond) {
+		js.Status.MarkJobSucceeded()
+	} else if cond != nil && jobConditionFailed(cond) {
+		js.Status.MarkJobFailed(cond.Reason, cond.Message)
+	}
+
+	return nil
+}
+
+// reconcileSink attempts to reconcile the sink object reference to a URI and set the sink in the JobSourceStatus.
+func (r *Reconciler) reconcileSink(ctx context.Context, js *v1alpha1.JobSource) error {
+	if js.Spec.Sink == nil {
+		js.Status.MarkNoSink("Missing", "Sink missing from spec")
+		return errSinkMissing
+	}
+
+	ref := js.Spec.Sink
+	if ref.Namespace == "" {
+		ref.Namespace = js.Namespace
+	}
+
+	desc := fmt.Sprintf("%s/%s, %s", js.Namespace, js.Name, js.GroupVersionKind().String())
+	uri, err := r.sinkReconciler.GetSinkURI(ref, js, desc)
+	if err != nil {
+		js.Status.MarkNoSink("NotFound", "Could not get sink URI from %s/%s: %v", ref.Namespace, ref.Name, err)
+		return err
+	}
+
+	if parsedURI, err := url.Parse(uri); err != nil || !isGoodURL(parsedURI) {
+		js.Status.MarkNoSink("InvalidValue", "SinkURI resolved to invalid sink: %s", uri)
+		return errSinkInvalid
+	}
+
+	js.Status.MarkSink(uri)
+
+	return nil
+}
+
+func (r *Reconciler) getJob(ctx context.Context, owner metav1.Object, ls labels.Selector) (*batchv1.Job, error) {
+	list, err := r.KubeClientSet.BatchV1().Jobs(owner.GetNamespace()).List(metav1.ListOptions{
```

Comment:
We should rewrite this to have deterministic names so we do not have to list and filter.

This comment is to mark the spot and make an issue.

Reply:
Done